### PR TITLE
C51-37: Making UI full bleed sass integration with gulp and add stylelint

### PR DESCRIPTION
**Pull request by swastikpareek**
## Overview
 PR covers following features

* Setup Folder structure for SASS.
* Setting up Gulp task for CSS complilation from SASS and adding minification to it
* Adding stylelint linter for scss check
* Make the UI for civicase to take full width

## Before
c51-37 - before

## After
c51-37 - after

## Technical Details
 * Setup Folder structure for SASS.
  ```
      scss
        |- base (Contains base code)
        |  |- mixins (Contains partials for mixins)
        |  |- typography (contains partials for app typography)
        |  |- variables (contains partials for variables used sitewide)
        |
        |- components (partials for different components)
        |- overrides (overrides if any) 
        |- pages (Partials specific to pages) 
  ```
  c51-37 - sass structure

* Replace occurence of `civicase-main` class to `civicase__container` making the HTML & CSS more BEM oriented.
* Gulp is configured to add `#bootstrap-theme` selector to each css styles except body classes and `.___outside-namespace` class.
*  The sourcemap file is generated everytime sass is compiled and compressed.
* Resets margin for `#page` element inside `page-civicrm-case-a` and remove max width of `1200px` from `#bootstrap-theme.civicase-main` selector to make the UI span 100%.
* For stylelint following rules are added in package.json
```
  "selector-class-pattern": null, // selector class pattern to follow - more info here - https://stylelint.io/user-guide/rules/selector-class-pattern/
  "max-nesting-depth": 3, // max depth which the scss should be nested
  "no-extra-semicolons": true, //outputs error if extra semicolon are present
  "no-duplicate-selectors": true, // flags duplicate selectors
  "scss/dollar-variable-colon-space-after": "at-least-one-space" // number of space to be added after colon - more info here - https://github.com/kristerkari/stylelint-scss/issues/195
```
* Global variables defined in package.json
 ** CRM
 ** angular

### CODE REVIEW

**Comment by deb1990** [comment]
@swastikpareek the technical details should be moved to Technical Details section



## Overview
 PR covers following features

 * Setup Folder structure for SASS.
  * Setting up Gulp task for CSS complilation from SASS and adding minification to it 
  * Adding stylelint linter for scss check
  * Make the UI for civicase to take full width

## Technical Details
  * add folder structure here.
  * add the details about the civicase-main class to civicase__container class here
  * Gulp is configured to add `#bootstrap-theme` selector to each css styles except body classes and `.___outside-namespace` class.
  *  The sourcemap file is generated everytime sass is compiled and compressed.

File: gulpfile.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+const postcss = require('gulp-postcss');
+const postcssPrefix = require('postcss-prefix-selector');
+const postcssDiscardDuplicates = require('postcss-discard-duplicates');
+const stripCssComments = require('gulp-strip-css-comments');
+const transformSelectors = require('gulp-transform-selectors');
+const cssmin = require('gulp-cssmin');
+const rename = require('gulp-rename');
+const sourcemaps = require('gulp-sourcemaps');
+
+// Local variables
```
Comment (deb1990) [inline gulpfile.js]: All variables in this files are local. So this comment does not really make sense. I would say no need to comment `// Local variables` and `// Gulp related variables`, its quite evident by looking at them. An empty line between them is enough(as you have it).

File: scss/pages/_pages.scss [hunk 2]
```diff
@@ -0,0 +1,5 @@
+.page-civicrm-case-a {
```
Comment (deb1990) [inline scss/pages/_pages.scss]: This change is not described in the PR docs. Please add details.

File: gulpfile.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  */
+gulp.task('sass', () => {
+  return gulp.src('scss/civicase.scss')
+    .pipe(bulk())
+    .pipe(sourcemaps.init())
+    .pipe(sass({
+      outputStyle: 'compressed',
+      precision: 10
+    }).on('error', sass.logError))
+    .pipe(stripCssComments({ preserve: false }))
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply gulpfile.js]: https://github.com/sindresorhus/strip-css-comments#preserve.
This strips all comments.

File: gulpfile.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+const cssmin = require('gulp-cssmin');
+const rename = require('gulp-rename');
+const sourcemaps = require('gulp-sourcemaps');
+
+// Local variables
+const bootstrapNamespace = '#bootstrap-theme';
+const outsideNamespaceRegExp = /^\.___outside-namespace/;
+
+/**
+  * The gulp task compiles civicase.scss file to civicase.min.css (minified file).
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply gulpfile.js]: Could you please explain what is to be done here?

File: gulpfile.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  * Tasks
+  * default : Runs SASS task
+  * sass:sync : Runs sass sync task for civicrmScssRoot
+  * sass: Compiles civicase.scss under scss folder to CSS counterpart
+  * watch: Watches for scss file changes and run sass task
+  */
+
+// Gulp related variables
+const gulp = require('gulp');
+const bulk = require('gulp-sass-bulk-import');
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply gulpfile.js]: We need this for importing SCSS directories in the file.
https://www.npmjs.com/package/gulp-sass-bulk-import

File: scss/base/typography/readme.md [hunk 6]
```diff
@@ -0,0 +1,2 @@
+### Folder details
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply scss/base/typography/readme.md]: Added.

File: scss/civicase.scss [hunk 7]
```diff
@@ -0,0 +1,12 @@
+/**
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply scss/civicase.scss]: Done. Please check

File: scss/civicase.scss [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+  * @file
+  * The file contains all the scss imports required for the civicase extensions
+  * All basic components like variables, mixins, typographys are defined first
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply scss/civicase.scss]: Updated.

File: templates/CRM/Civicase/Page/ContactActivityTab.tpl [hunk 9]
```diff
@@ -20,4 +20,4 @@
     angular.bootstrap(document.getElementById('civicaseActivitiesTab'), ['civicaseActivitiesTab']);
   });
 </script>
-{/literal}
\ No newline at end of file
+{/literal}
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply templates/CRM/Civicase/Page/ContactActivityTab.tpl]: There was an empty line missing in the last (Semistandard rule), which have been added here.

File: gulpfile.js [hunk 10]
```diff
@@ -0,0 +1,78 @@
+'use strict';
+
+/**
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply gulpfile.js]: Done. Please check

File: scss/overrides/readme.md [hunk 11]
```diff
@@ -0,0 +1,2 @@
+### Folder details
+This folder contains partials for overrides (if needed).
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply scss/overrides/readme.md]: done

File: scss/pages/readme.md [hunk 12]
```diff
@@ -0,0 +1,2 @@
+### Folder details
+This folder contains partials for page related sass
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply scss/pages/readme.md]: done

File: gulpfile.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (swastikpareek) [reply gulpfile.js]: Fixed.

File: scss/pages/_pages.scss (hunk 2 above)
Comment (swastikpareek) [reply scss/pages/_pages.scss]: Added.

File: gulpfile.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (swastikpareek) [reply gulpfile.js]: done.

File: gulpfile.js [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+  * @file
+  * This file contains gulp configurations for setting up SASS with feature of
+  * importing Shoreditch Partials and minifying the css file to .min.css
+  * Tasks
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply gulpfile.js]: Added.

File: package.json [hunk 14]
```diff
@@ -15,8 +15,33 @@
...
     ]
   },
+  "stylelint": {
+    "extends": "stylelint-config-sass-guidelines",
+    "plugins": [
+      "stylelint-order",
+      "stylelint-scss"
+    ],
+    "rules": {
+      "selector-class-pattern": null,
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply package.json]: Done.

File: package.json [hunk 15]
```diff
@@ -15,8 +15,33 @@
...
       "semistandard --fix",
       "git add"
+    ],
+    "*.scss": [
+      "stylelint --fix",
+      "git add"
+    ]
+  },
+  "semistandard": {
+    "globals": [
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply package.json]: Done.

File: gulpfile.js [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+ * @file
+ * This file contains gulp configurations for setting up SASS with feature of
+ * importing Shoreditch Partials and minifying the css file to .min.css
+ *
+ * Tasks
+ * default : Runs SASS task
+ * sass: Compiles civicase.scss under scss folder to CSS counterpart
+ * watch: Watches for scss file changes and run sass task
+ */
+const gulp = require('gulp');
```
Comment (deb1990) [inline gulpfile.js]: Sorry I forgot to mention. Lets not use ES6 yet, as we are not using ES6 anywhere else. We definitely want to switch to ES6 in future, then we would switch both the angular and gulpfile. Till then lets use ES5 only. Please fix this for the whole file.

Comment (swastikpareek) [reply gulpfile.js]: Done.

**Pull request merged**